rfcs/77-dev-tools-abstraction.md

```diff
@@ -0,0 +1,658 @@
+            let Ok(command) = dev_command_metadata.from_str(&event.0) else {
+                warn!("Could not parse the command from the supplied string");
+                continue;
+            }
+
+            // Now we can run the command directly on the `World` using `Command::apply`!
+            command.apply(world);
+        }
+    })
+}
+```
+
+While a number of other features could sensibly be added to this API (a `--help` flag, saving and loading config to disk, managing compatibility between dev tools),
+this MVP should be sufficient to prove out the viability of the core architecture.
+
+## Implementation strategy
+
+### What metadata do toolboxes need?
+
+We can start simple:
+
+- the type name of the dev tool
+- any doc strings on the dev tool struct
+- the value and types of any configuration that must be supplied
+
+By relying on simple structs to configure both our modal dev tools and commands, we can extract this information automatically, via reflection.
+Additional dev tool specific metadata (such as a classification scheme) can be added to the traits on an opt-in basis in the future.
+
+Optional methods on both `ModalDevTool` and `DevCommand` will allow us to override the supplied defaults if needed.
+
+We also need access to one other critical piece of information: a function pointer that allows us to construct a new value of this type from a string.
+
+As a result our `DevToolMetadata` looks like:
+
+```rust
+struct DevToolMetadata {
+   name: String,
+   type_info: TypeInfo,
+   from_str: StringConstructorFn,
+}
+```
+
+The tricky bit comes when we get to `StringConstructorFn`: we need to be able to store a function pointer that will take us from a string, and return an object that implements our core traits.
+
+TODO: how can this be done?
+
+The definition for `DevCommandMetadata` is effectively identical to start, but over time we should expect these to diverge: splitting them from the beginning will ease migrations going forward.
+
+### What do the registries for our dev tools look like?
+
+Keeping track of the registered dev tools without storing them all in a dedicated collection is quite challenging!
+How can we keep track of it under the hood?
+
+```rust
+#[derive(Resource)]
+struct DevToolsRegistry {
+    /// The stored collection of modal dev tools, tracked in a type-erased way using [`ComponentId`]
+    /// 
+    /// The key is the `name()` provided by the `ModalDevTool` trait.
+    modal_dev_tools: HashMap<String, ToolMetaData>,
+    /// The metadata for all registered dev commands.
+    /// 
+    /// The key is the `name()` provided by the `DevCommand` trait.
+    dev_commands: HashMap<String, DevCommandMetaData> 
+}
+```
+
+There are a few key operations that we will need to perform:
+
+```rust
+impl DevToolsRegistry {
+    /// Gets a reference to the specified modal dev tool by `ComponentId`.
+    fn get_tool_by_id(world: &World, component_id: ComponentId) -> Option<&dyn ModalDevTool> {
```

Comment:
As said above, we can't have this as object safe, but maybe we can have this return dyn Reflect?